torch/_higher_order_ops/auto_functionalize.py

```diff
@@ -120,44 +123,185 @@ def can_auto_functionalize(op: OperatorBase) -> bool:
     return True
 
 
+@dataclass
+class ViewInfo:
+    base_index: int
+    size: Any = None
+    stride: Any = None
+    storage_offset: Any = None
+    # When is_view is false, the tensor is the base.
+    # When is_view is true, size, stride and storage_offset are all None.
+    is_view: bool = True
+
+    def regenerate_view(self, bases_list: List[Tensor]):
+        if not self.is_view:
+            return bases_list[self.base_index]
+        return torch.as_strided(
+            bases_list[self.base_index],
+            self.size,
+            self.stride,
+            self.storage_offset,
+        )
+
+
+# arg_to_base_index maps arg_name to ViewInfo | [ViewInfo]
+def write_view_information_to_args(
+    arg_names: List[str],
+    arg_types: List[torch.Type],
+    input_kwargs: Dict[str, Any],
```

Comment:
Nit: this might be cleaner if we accepted mutated_arg_names, mutated_arg_types, and mutated_arg_nodes instead of (arg_names, arg_types, input_kwargs). mutated_arg_nodes would just be [input_kwargs[name] for name in arg_names]

Reply:
i only bass kwargs now (which is the arg of the custom op added doc, no need for both